jpa/subsystem/src/main/java/org/jboss/as/jpa/transaction/TransactionUtil.java

```diff
@@ -137,47 +181,50 @@ public SessionSynchronization(EntityManager session, String scopedPuName) {
         }
 
         public void beforeCompletion() {
+            afterCompletionCalled = false;
         }
 
-        public void afterCompletion(int status) {
+        public synchronized void afterCompletion(int status) {
             /**
-             * If its not safe (safeToClose returns false) to close the EntityManager now,
-             * any connections joined to the JTA transaction
-             * will be released by the JCA connection pool manager.  When the JTA Transaction is no longer
-             * referencing the EntityManager, it will be eligible for garbage collection.
-             * See AS7-6586 for more details.
+             * Note: synchronization is to protect against two concurrent threads from closing the EntityManager (manager)
```

Comment:
Under what condition is there two threads calling afterCompletion? Shouldn't there only be one transaction per synchronization? Otherwise that implies there are two threads operating on an entity manager, and those aren't thread safe.